### ISSUE TITLE
Discussion: Would we miss the robocopy functionality in xSQLServerSetup?

### ISSUE THREAD

**Issue by johlju** [PowerShell/xSQLServer#167]
I would like to raise this question to see if there is a scenario where someone would like to copy the installation files from the source to the server being installed. 
Currently this is done when the parameter `SourceCredential` is assigned a `PSCredential`  objekt.

```PowerShell
    if($SourceCredential)
    {
        NetUse -SourcePath $SourcePath -Credential $SourceCredential -Ensure "Present"
        $TempFolder = [IO.Path]::GetTempPath()
        & robocopy.exe (Join-Path -Path $SourcePath -ChildPath $SourceFolder) (Join-Path -Path $TempFolder -ChildPath $SourceFolder) /e
        $SourcePath = $TempFolder
        NetUse -SourcePath $SourcePath -Credential $SourceCredential -Ensure "Absent"
    }
```

If we want to keep this I have to change the code so I'm able to mock this in the tests I'm writing. If this is something that is not necessary, then I can ignore this in the test and we could remove the functionality. Removing it would be a breaking change.

**Comment by johlju** [comment]
I had some thought when writing this issue why it would be good to remove the functionality.
* Takes up disk space on the server being configured.
* Another DSC resource could move files in and between the file systems, and do it better. xSQLServerSetup could be dependent on that resource in the config file.

But on the other hand if the functionality was remove, we would could lose some things too.
* It could be useful to have the actually media bits on the server to know what bits was used to installed the 
* If the bits is on the local server. Manually changes could be made easier thru GUI.

**Comment by luigilink** [comment]
I think we can use "file" dscresource like a prerequesites of SQL installation.
We have the same probleme with a lot of DSC modules, before using the setup.exe, you have to copy/paste the binaries in a local folder.
And for the disk space, I don't know if a dsc resource already exists?

```
File APPLICATION_SQLGetSources
{
      SourcePath = ""
      DestinationPath = ""
      Type = "Directory"
      Recurse = $true
      PsDscRunAsCredential = (if needed)
}
```

**Comment by johlju** [comment]
Great suggestion! I will look at the File resource and see if that can replace this.
 
One other thing on the bad side of copy everything. Install takes longer. First move all the bits over the network, then installing. Should save some time to just install directly over network. 
On the other hand, if installing over a network that has network glitches could be bad. Then it will be good to use the File resource if it works.

I also see now that the Get-method will always use the source, never the local copy of the bits. 
```PowerShell
    if($SourceCredential)
    {
        NetUse -SourcePath $SourcePath -Credential $SourceCredential -Ensure "Present"
    }
    $Path = Join-Path -Path (Join-Path -Path $SourcePath -ChildPath $SourceFolder) -ChildPath "setup.exe"
    $Path = ResolvePath $Path
    Write-Verbose "Path: $Path"
    $SQLVersion = GetSQLVersion -Path $Path
```

**Comment by nabrond** [comment]
With the File resource approach, the installation media would always have to be present on the server. If you removed it, DSC would just recopy the bits. The robocopy approach at least puts the files in a temporary location that can be cleaned up later to preserve space on the node.

The Get method has no need to actually copy the media to the node, it only uses "setup.exe" to determine the version of the bits to be installed.

**Comment by johlju** [comment]
@nabrond That is true. That takes File resource out of the picture.
Regarding the Get-method using setup.exe on the source. That means that the Source must be present on ever pull/push for the Test-method to work. It could as easily install from the source as well :)

It seems it might still be a need for the robocopy functionality?

**Comment by nabrond** [comment]
You are correct @johlju, Source must always exist for Get and Test to work. Maybe Get needs to be updated to determine the installed SQL version some other way? It could still rely on the setup.exe version for Set. 
Regardless, I think there are valid use cases for being able to install from a local copy of the media. Maybe this should be another option?

**Comment by johlju** [comment]
We keep it in then, until we get a better reason to not keep it in.

In the Get-method the major version number is used in four instances (`$SQLVersion` local variable).

 * Building the service name of the Integration Service. Ie. `MsDtsServer120`. This is later used to determine if the service installed or not (and adds 'IS' to the features list).
 * Used to determine if SSMS is installed. Each major version has different unique product identifiers. 
 * Used to determine if ADV_SSMS is installed. Each major version has different unique product identifiers. 
* Getting the paths to the Shared Directory and Shared WOW Directory.

We need to have another way to individually get the correct major version number for these, since each can be installed individually. I think the original coder used the setup.exe to know exactly what version was to be evaluated.

**Comment by nabrond** [comment]
I like the idea of not relying on setup.exe for determining version information in the Get method. This should be fairly easy to get from the registry at ````HKLM:\Software\Microsoft\Microsoft SQL Server\<Instance>\MSSQLServer\CurrentVersion```` since the method would know the instance that should be installed.

**Comment by jackgfdreacher** [comment]
NICE ARTICLE

**Comment by hemant51** [comment]
use robocopy

**Comment by georgeeerenee** [comment]
We have the same probleme with a lot of DSC modules, before using the setup.exe, you have to copy/paste the binaries in a local folder.
And for the disk space, I don't know if a dsc resource already exists?...it's better to use an alternative tool of it like G.S Richcopy 360 Enterprise because i already used robocopy and find some issues in it.

**Comment by georgeeerenee** [comment]
We have the same probleme with a lot of DSC modules, before using the setup.exe, you have to copy/paste the binaries in a local folder.
And for the disk space, I don't know if a dsc resource already exists?...it's better to use an alternative tool of it like G.S Richcopy 360 Enterprise because i already used robocopy and find some issues in it.

**Comment by johlju** [comment]
@georgeeerenee current implementation in SqlSetup only uses `robocopy` if the parameter `SourceCredential` is assigned credentials. So if you use another tool to copy the media to the target node, then you just need to specify the local path in the parameter `SourcePath` and make sure either the SYSTEM account or the credentials specified in the parameter  `PSDscRunAsCredential` have permission to the local folder.
